eyeshade/controllers/referrals.js

```diff
@@ -1,23 +1,68 @@
 const boom = require('boom')
 const bson = require('bson')
+const BigNumber = require('bignumber.js')
 const Joi = require('@hapi/joi')
 const underscore = require('underscore')
+const _ = underscore
 
 const utils = require('bat-utils')
 const braveHapi = utils.extras.hapi
 const braveJoi = utils.extras.joi
+const extrasUtils = utils.extras.utils
+
+const queries = require('../lib/queries')
 
 const v1 = {}
 
-const schema = Joi.array().min(1).items(Joi.object().keys({
+const amountValidator = braveJoi.string().numeric()
+const groupNameValidator = Joi.string().optional().description('the name given to the group')
+const channelIdValidator = braveJoi.string().publisher().required().description('the publisher identity')
+const currencyValidator = braveJoi.string().altcurrencyCode().description('the currency unit being paid out')
+const groupIdValidator = Joi.string().guid().required().description('the region from which this referral came')
+const countryCodeValidator = braveJoi.string().countryCode().allow('OT').description('a country code in iso 3166 format').example('CA')
+const referral = Joi.object().keys({
   ownerId: braveJoi.string().owner().required().description('the owner'),
-  channelId: braveJoi.string().publisher().required().description('the publisher identity'),
+  channelId: channelIdValidator,
   downloadId: Joi.string().guid().required().description('the download identity'),
   platform: Joi.string().token().required().description('the download platform'),
   finalized: Joi.date().iso().required().description('timestamp in ISO 8601 format').example('2018-03-22T23:26:01.234Z')
-}).unknown(true)).required().description('list of finalized referrals')
-
-let altcurrency
+})
+const manyReferrals = Joi.array().min(1).items(referral).required().description('list of finalized referrals')
+const groupStampedReferral = referral.keys({
+  downloadTimestamp: Joi.date().iso().required().description('the timestamp when the referral was downloaded to apply correct payout to it'),
+  groupId: groupIdValidator
+})
+const manyGroupStampedReferrals = Joi.array().min(1).items(groupStampedReferral).required().description('list of finalized referrals to be shown to publishers')
+
+const anyReferralVersion = Joi.alternatives().try(
+  manyGroupStampedReferrals,
+  manyReferrals
+)
+
+const referralGroupCountriesValidator = Joi.object().keys({
+  id: Joi.string().guid().required().description('the group id to report back for correct value categorization'),
+  activeAt: Joi.date().iso().required().description('the download cut off time to honor the amount'),
+  name: groupNameValidator.optional().description('name of the group'),
```

Comment:
I would expect a name would be required for a group as that's what we're likely going to display to the end user

Reply:
so, this is a fair point, but ledger is serving multiple services with this endpoint. it may make more sense to make activeAt optional, since everyone needs the id and the cross section of needs of fields is / will be different for everyone